### PR TITLE
Landmark data preparation

### PR DESCRIPTION
**Pull request by alainamstutz**
Hi Venexia,
I would love to have your review. It contains:
data_process.R, which does the following:
1. Import/extract the feather dataset from the dataset definition that Zoe already reviewed and approved
2. Basic type formatting of variables -> fn_extract_data.R()
3. Process some covariates and apply the diabetes algorithm -> fn_diabetes_algorithm()
4. Evaluate/apply the quality criteria -> fn_quality_assurance_midpoint6()
5. Evaluate/apply the completeness criteria: -> fn_completeness_criteria_midpoint6()
6. Evaluate/apply the eligibility criteria: -> fn_elig_criteria_midpoint6() & fn_apply_elig_criteria()

The covariates are "attached" to the landmark date ("landmark_date"), while the eligibility criteria are "attached" to the individual incident T2DM diagnosis date ("elig_date_t2dm")

I then realized that by applying the quality & completeness criteria, it reduced my (dummy) dataset substantially and thus was worried about the numbers if we only stick to the period of incident T2DM between mid2018 and mid2019. Therefore I extended the eligibility yearly until mid2013, added the treatment/exposure and main outcome variables, and exported all the 6 csv files containing the most important numbers re feasibility - just to double-check. 

Along the way I made a few improvements to the dataset definition (already reviewed by Zoe), but these are only annotations, 1 error about sex, and streamlining variable names for the diabetes algo. 

Happy to discuss the code next week briefly face-to-face if you think that makes sense. I also have a couple of questions e.g. regarding importing the metadates. 

Thanks a lot!
Best,
-Alain

### CODE REVIEW

File: analysis/data_process.R
Comment (pmadleydowd) [reply analysis/data_process.R:1]: Running line 137 i get an error when running interactively or via opensafely run:

Error in `mutate()`:
ℹ In argument: `tmp_elig_year_first_diabetes_diag = as.integer(format(tmp_elig_date_first_diabetes_diag, "%Y"))`.
Caused by error:
! object 'tmp_elig_date_first_diabetes_diag' not found
Run `rlang::last_trace()` to see where the error occurred.

Please can you try to fix this before I go further?

**Review by pmadleydowd — COMMENTED** [review]
Have hit a roadblock - please see comment on data_process.R

File: analysis/data_process.R
Comment (alainamstutz) [reply analysis/data_process.R:1]: Ah, I see what the problem is. After creating the PR, I then went ahead and made some very minor adaptation to the dataset definition (which was already reviewed and approved by Zoe) because I realized that the naming of the variable was inconsistent and changed the name from tmp_cov_date_first_diabetes_diag to tmp_elig_date_first_diabetes_diag. That's why it works all fine on my side (just tested again) and on GitHub, but not in your current local repo since you would first have to update the dataset definition to the new variable name through this individual commit: https://github.com/opensafely/metformin_covid/pull/9/commits/1371f4a6adddafd4c7838be13388a7387d49ab6b

Sorry, this was not a smart decision, should have opened a new branch for that. How do we solve this? can you download the latest and correct "dataset_definition_landmark.py" manually (https://github.com/opensafely/metformin_covid/blob/1371f4a6adddafd4c7838be13388a7387d49ab6b/analysis/dataset_definition_landmark.py), replace this file in your local repo and then run it locally?

Comment (pmadleydowd) [reply analysis/data_process.R:1]: Implemented and is now working

Comment (pmadleydowd) [reply analysis/data_process.R:1]: - Section 1 and 2 - all ok 
- Section 3 define covariates and apply diabetes algorithm (see comments on diabetes algorithm) all seems ok 
- Section 4 apply quality criteria ...

File: analysis/functions/eligibility_midpoint6.R
Comment (pmadleydowd) [reply analysis/functions/eligibility_midpoint6.R:1]: # fn_elig_criteria_midpoint
- crit1 - unclear to me if the function allows for flexibility in the year used as the filtering is done on mid2018_date. 
- crit2-crit6 - seem ok
- counts after exclusion - this bit seems like double programming again to me

# fn_apply_elig_criteria 
- double programming of previous function

**Review by pmadleydowd — CHANGES_REQUESTED** [review]

File: analysis/functions/fn_diabetes_algorithm.R
Comment (alainamstutz) [reply analysis/functions/fn_diabetes_algorithm.R:1]: Very good points - thanks a lot! 
Re algorithm: https://docs.google.com/document/d/1Vptfzx-S981m7Th1FIDbAjwe1iUchrXFh0Y6HrVzrDQ/edit?tab=t.0#heading=h.66b485waohz1
1) According to algorithm, all codes are searched in both sources (primary and secondary care) and used at all steps, except step 6a and 6b, where only primary care codes are used. Added a footnote. However, I need to double-check with the post covid diabetes team, it seems that gestational DM codes are only searched in primary care...?! Added a note to the protocol.
2) Added a footnote to algorithm
3) No, not possible according to ehrQL count_for_patient() function: https://docs.opensafely.org/ehrql/reference/language/#PatientFrame.count_for_patient. Added a note in the R script
4) Thanks for reminding! I had made a note there, but forgot about it again. I think in general Type 1 codes overrule Type 2 codes (that's what I at least would make sense from a clinical perspective), but I will double-check with the post covid diabetes team. Added a note to the protocol.
5) 6.5% = 47.5 mmol/mol. Added a footnote to the algorithm.

File: analysis/functions/fn_completeness_criteria_midpoint6.R
Comment (alainamstutz) [reply analysis/functions/fn_completeness_criteria_midpoint6.R:1]: Rule 3 and 4: I blindly took this over from the post covid diabetes team, very valid point, esp. NHS establishment :-) I changed it to Rule 3 = 1948 and Rule 4 = studied_date, but added a comment to the protocol to double-check

File: analysis/data_process.R
Comment (alainamstutz) [reply analysis/data_process.R:1]: Re "arguments section": Something I still don't grasp 100%. Isn't this for the analysis script when you want to run several analyses in a loop (i.e. primary model, secondary, subgroup, sensitivity, etc.)? not sure I understand what "interactively" means, since it works well locally, on GitHub and on Jobs (where I did run it for the feasibility numbers). Definitely need to understand this better.

Section 3: re diabetes algo: See answers above

Section 4: Thanks for picking up the mix-up between completeness criteria and quality assurance criteria, the titles were wrong in the protocol, I adapted. Really really good suggestion re double/triple coding - I implemented it, I think this is much smarter, shorter, more efficient, indeed. Thanks a lot.

Section 5: Applied same structure as for Section 4. The rule for "Known age" is now implemented (thanks!). The "bin_was_registered" was correct, I went back 30 months (914 days) -> practice_registrations.spanning(landmark_date - days(914), landmark_date). But actually, your comment made me think again of this criteria, and knowing now from the workshop how it is implemented in ehrQL, I think 1 year is sufficient. I changed it to 1 year in protocol and dataset_definition. 

Section 6: Applied same structure as for Section 4 and changed to 1 function. Criteria 1 should filter those with no T2DM or T2DM outside of the applied window: a) too close to landmark date = within previous 6 months / or b) too old starting with "before mid2018" (mid2018_date - years_in_days = 0), then going back in yearly steps until mid2013 in a loop.

File: analysis/functions/fn_elig_criteria_midpoint6.R
Comment (pmadleydowd) [inline analysis/functions/fn_elig_criteria_midpoint6.R:1]: matches protocol  and implements single coding (double coding removed)

minor comment: for clinical history variables do you need to consider if the exclusion criteria occurs on the same day that type 2 diabetes diagnosis

Waiting on response from Alain before approval

**Review by pmadleydowd — COMMENTED** [review]
few more checks needed before approval - have stated which files approved and which need more comments in the comments section of each file

File: analysis/functions/fn_quality_assurance_midpoint6.R
Comment (alainamstutz) [reply analysis/functions/fn_quality_assurance_midpoint6.R:1]: Re 1793: It is from the post covid diabetes repo. 1948 would not be a good idea, then we exclude 70+, so 1793 makes somehow sense (but not related to NHS establishment). I have not received an answer yet, but I added the query in my protocol and will not forget it: https://docs.google.com/document/d/1Vptfzx-S981m7Th1FIDbAjwe1iUchrXFh0Y6HrVzrDQ/edit?disco=AAABZ2enM_k

Re Sys.Date() for the Rule 4: Makes sense to me, would keep it. Added a comment to my protocol: https://docs.google.com/document/d/1Vptfzx-S981m7Th1FIDbAjwe1iUchrXFh0Y6HrVzrDQ/edit?disco=AAABaBF8DkU

File: analysis/data_process.R
Comment (alainamstutz) [reply analysis/data_process.R:1]: Hi Paul, based on discussions with Venexia and Rachel on Thursday, I added treatment patterns to the feasibility data output, since we are not sure what the best design is (this landmark design is maybe also not really the fitting set up..). So, unfortunately still lots of exploring. I think for exploration out_date_covid19_severe is fine (just to know how many events after landmark/pandemic start). You'll now see in the updated data_process that I a) added treatment patterns and b) modified the fn_elig_criteria_midpoint6 to re-apply the eligibility criteria again at landmark (also based on discussion from Thursday). 
I am fully aware that I am now just adding stuff to this file - and onwards to you. So, I am also happy to reassign this file to someone else!

Comment (pmadleydowd) [reply analysis/data_process.R:1]: Sorry I wont have time to review any more changes this week. Happy to do further review in Feb 2025 once back

Comment (alainamstutz) [reply analysis/data_process.R:1]: I fully understand!

**Pull request merged**
